### PR TITLE
Move LLPC timer profile related code to utility class TimerProfiler

### PR DESCRIPTION
**Pull request by trenouf**
TimerProfiler includes two timer groups
- Total Group: total time for all operations, it is equal with the lift time of TimerProfiler
- Phase Group: time for each phase, it can be enabled by enableMask in construct function.

Beside these, all timers in TimerProfiler are controlled by llvm flag TimePassesIsEnabled

Change-Id: I646445ef414e2c10043353abd6be860659088cdb

### CODE REVIEW

**Comment by trenouf** [comment]
This is a modified version of Qun's #238.

**Comment by amdvlk-admin** [comment]
### Test summary for commit ad513f3e1185d0f02273a48ea5637020a92bb387
<details open><summary>Driver commits used in build</summary><ul>

<li>CWPACK: amd-master b601c88aeca7a7b08becb3d32709de383c8ee428</li>
<li>METROHASH: amd-master 2b6fee002db6cc92345b02aeee963ebaaf4c0e2f</li>
<li>PAL: dev 88d997710b4e405f3a8e3fd60a38afee9e3e77e2</li>
<li>SPVGEN: dev 2f31d1170e8a12a66168b23235638c4bbc43ecdc</li>
<li>XGL: dev 2cb5558b94c5dc839e093cb439057a1802426c8e</li>
<li>LLVM: amd-vulkan-dev dcb42575c3d374467bb6e4e609f8bf6d1302c297</li>
</ul></details>

<details open><summary>ShaderDB test</summary><ul>

<li>Passed: 927/927 (100.0%)</li>
<li>Failed: 0/927 (0.0%)</li>
</ul></details>

<details open><summary>CTS tests (built with branch Vulkan-CTS-1.1.2.3-cherry-picks)</summary><ul>


<details open><summary>Ubuntu 18.04, Ellesmere</summary><ul>

<li>Passed: 68331/68331 (100.0%)</li>
<li>Failed: 0/68331 (0.0%)</li>
</ul></details>

<details open><summary>Rhel 7.6, Vega10</summary><ul>

<li>Passed: 68363/68363 (100.0%)</li>
<li>Failed: 0/68363 (0.0%)</li>
</ul></details>


</ul></details>

**Comment by amdrexu** [comment]
I will let Qun make comment since it revises his change.

File: util/llpcTimerProfiler.cpp [hunk 1]
```diff
@@ -0,0 +1,191 @@
...
+    {
+        // Stop whole timer
+        m_wholeTimer.stopTimer();
+    }
+}
+
+// =====================================================================================================================
+// Get a specific timer. Returns nullptr if !TimePassesIsEnabled.
+Timer* TimerProfiler::GetTimer(
+    TimerKind    name)           // Kind of phase timer
```
Comment (amdrexu) [inline util/llpcTimerProfiler.cpp]: name -> timerKind

File: util/llpcTimerProfiler.cpp [hunk 2]
```diff
@@ -0,0 +1,191 @@
...
+{
+    return TimePassesIsEnabled ? &m_phaseTimers[name] : nullptr;
+}
+
+// =====================================================================================================================
+// Adds pass to start or stop timer in PassManager
+void TimerProfiler::AddTimerStartStopPass(
+    legacy::PassManager*  pPassMgr,       // Pass Manager
+    TimerKind             name,           // Kind of phase timer
```
Comment (amdrexu) [inline util/llpcTimerProfiler.cpp]: name -> timerKind

File: util/llpcTimerProfiler.cpp [hunk 3]
```diff
@@ -0,0 +1,191 @@
...
+    if (TimePassesIsEnabled)
+    {
+        pPassMgr->add(CreateStartStopTimer(&m_phaseTimers[name], start));
+    }
+}
+
+// =====================================================================================================================
+// Starts or stop specified timer.
+void TimerProfiler::StartStopTimer(
+    TimerKind name,         // Kind of phase timer
```
Comment (amdrexu) [inline util/llpcTimerProfiler.cpp]: name -> timerKind

File: util/llpcTimerProfiler.h [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+public:
+    TimerProfiler(uint64_t hash64, const char* pDescriptionPrefix, uint32_t enableMask);
+
+    ~TimerProfiler();
+
+    Timer* GetTimer(TimerKind name);
+
+    void AddTimerStartStopPass(legacy::PassManager* pPassMgr, TimerKind name, bool start);
+
+    void StartStopTimer(TimerKind name, bool start);
```
Comment (amdrexu) [inline util/llpcTimerProfiler.h]: name -> timerKind

File: util/llpcTimerProfiler.h [hunk 5]
```diff
@@ -0,0 +1,103 @@
...
+// =====================================================================================================================
+// Represents a utility class for time profile, it wraps LLVM Timer and TimerGroup in internal.
+class TimerProfiler
+{
+public:
+    TimerProfiler(uint64_t hash64, const char* pDescriptionPrefix, uint32_t enableMask);
+
+    ~TimerProfiler();
+
+    Timer* GetTimer(TimerKind name);
```
Comment (amdrexu) [inline util/llpcTimerProfiler.h]: name -> timerKind

File: util/llpcTimerProfiler.h [hunk 6]
```diff
@@ -0,0 +1,103 @@
...
+
+    void AddTimerStartStopPass(legacy::PassManager* pPassMgr, TimerKind name, bool start);
```
Comment (amdrexu) [inline util/llpcTimerProfiler.h]: name -> timerKind

File: util/llpcTimerProfiler.h [hunk 7]
```diff
@@ -0,0 +1,103 @@
...
+#include "llpcDebug.h"
+
+namespace llvm
+{
+namespace legacy
+{
+
+class PassManager;
+
+}
```
Comment (amdrexu) [inline util/llpcTimerProfiler.h]: } -> } // legacy

File: util/llpcTimerProfiler.h [hunk 8]
```diff
@@ -0,0 +1,103 @@
...
+}
```
Comment (amdrexu) [inline util/llpcTimerProfiler.h]: } -> } // llvm

**Comment by amdrexu** [comment]
Qun has incorporarte your change to his one. If code conflict blocks your work, you can merge his change.

**Comment by trenouf** [comment]
Superseded by Qun modifying his #238 and it being merged.

**Pull request closed without merging**